rust/src/jitter.rs

```diff
@@ -0,0 +1,428 @@
+        }
+    }
+}
+
+pub struct Jitter<T: AccountProvider> {
+    drift_client: DriftClient<T>,
+    perp_params: DashMap<u16, JitParams>,
+    spot_params: DashMap<u16, JitParams>,
+    ongoing_auctions: DashMap<String, JoinHandle<()>>,
+    exclude_auction: Option<Box<ExcludeAuctionFn>>,
+    jitter: Arc<dyn JitterStrategy + Send + Sync>,
+}
+
+#[async_trait]
+pub trait JitterStrategy {
+    async fn try_fill(&self, taker: User, taker_key: Pubkey, taker_stats_key: Pubkey, order: Order, order_sig: String, referrer_info: Option<ReferrerInfo>, params: JitParams) -> JitResult<()>;
+}
+
+impl<T: AccountProvider + Clone> Jitter<T> {
+    pub fn new(
+        drift_client: DriftClient<T>,
+        jitter: Arc<dyn JitterStrategy + Send + Sync>, 
+    ) -> Arc<Self> {
+
+        Arc::new(Jitter {
+            drift_client,
+            perp_params: DashMap::new(),
+            spot_params: DashMap::new(),
+            ongoing_auctions: DashMap::new(),
+            exclude_auction: None,
+            jitter,
+        })
+    }
+
+    /// Set up a Jitter with the Shotgun strategy
+    pub fn new_with_shotgun(
+        drift_client: DriftClient<T>,
+        config: Option<RpcSendTransactionConfig>,
+        cu_params: Option<ComputeBudgetParams>,
+    ) -> Arc<Self> {
+        let jit_proxy_client = JitProxyClient::new(drift_client.clone(), config, cu_params);
+        let shotgun: Arc<dyn JitterStrategy + Send + Sync> = Arc::new(Shotgun { jit_proxy_client });
+
+        Arc::new(Jitter {
+            drift_client,
+            perp_params: DashMap::new(),
+            spot_params: DashMap::new(),
+            ongoing_auctions: DashMap::new(),
+            exclude_auction: None,
+            jitter: shotgun,
+        })
+    }
+
+    // Subscribe to auction events and start listening for them
+    pub async fn subscribe(self: Arc<Self>, url: String) -> JitResult<()>{
+        let (auction_sender, mut auction_receiver): (Sender<Box<dyn Event>>, Receiver<Box<dyn Event>>) = mpsc::channel(100); 
+
+        // Set up a receiver to bounce the auction events to the trampoline 
+        tokio::spawn(async move {
+            while let Some(event) = auction_receiver.recv().await {
+                let _ = self.on_auction_sync(event);
+            }
+        });
+
+        let auction_subscriber_config = AuctionSubscriberConfig {
+            commitment: CommitmentConfig::processed(),
+            resub_timeout_ms: None,
+            url,
+        };
+
+        let mut auction_subscriber = AuctionSubscriber::new(auction_subscriber_config);
+        auction_subscriber.subscribe().await?;
+
+        let auction_sender: Sender<Box<dyn Event>> = auction_sender.clone();
+
+        // Send auction events to the receiver to be processed
+        auction_subscriber.event_emitter.subscribe("auction",  move |event| {
+            let _ = auction_sender.try_send(event);
+        });
+
+        futures::future::pending::<()>().await;  // Keep the event loop alive forever
```

Comment:
not sure if this makes issues for killing the process. can try to select over `tokio::signal_ctrl_c` here too